airbyte-integrations/connectors/source-amazon-seller-partner/source_amazon_seller_partner/streams.py

```diff
@@ -102,79 +128,303 @@ def get_updated_state(self, current_stream_state: MutableMapping[str, Any], late
+
+    @staticmethod
+    def decrypt_aes(content, key, iv):
+        key = base64.b64decode(key)
+        iv = base64.b64decode(iv)
+        decrypter = AES.new(key, AES.MODE_CBC, iv)
+        decrypted = decrypter.decrypt(content)
+        padding_bytes = decrypted[-1]
+        return decrypted[:-padding_bytes]
+
+    def decrypt_report_document(self, url, initialization_vector, key, encryption_standard, payload):
+        """
+        Decrypts and unpacks a report document, currently AES encryption is implemented
+        """
+        if encryption_standard == "AES":
+            decrypted = self.decrypt_aes(requests.get(url).content, key, initialization_vector)
+            if "compressionAlgorithm" in payload:
+                return zlib.decompress(bytearray(decrypted), 15 + 32).decode("iso-8859-1")
+            return decrypted.decode("iso-8859-1")
+        raise Exception([{"message": "Only AES decryption is implemented."}])
+
+    def parse_response(self, response: requests.Response) -> Iterable[Mapping]:
+        payload = response.json().get(self.data_field, {})
+        document = self.decrypt_report_document(
+            payload.get("url"),
+            payload.get("encryptionDetails", {}).get("initializationVector"),
+            payload.get("encryptionDetails", {}).get("key"),
+            payload.get("encryptionDetails", {}).get("standard"),
+            payload,
+        )
 
-    def request_params(
-        self, stream_state: Mapping[str, Any], next_page_token: Mapping[str, Any] = None, **kwargs
-    ) -> MutableMapping[str, Any]:
-        params = super().request_params(stream_state, next_page_token, **kwargs)
-        if not next_page_token:
-            params.update({"reportTypes": self.name})
-        return params
+        document_records = csv.DictReader(StringIO(document), delimiter="\t")
+        yield from document_records
 
+    def read_records(self, *args, **kwargs) -> Iterable[Mapping[str, Any]]:
+        """
+        Create and retrieve the report.
+        Decrypt and parse the report is its fully proceed, then yield the report document records.
+        Yield the report body itself if there are no document id in it.
+        """
 
-class MerchantListingsReports(ReportsBase):
+        # create and retrieve the report
+        report_payload = self._retrieve_report()
+        is_done = report_payload.get("processingStatus") == "DONE"
+
+        if is_done:
+            # retrieve and decrypt the report document
+            document_id = report_payload["reportDocumentId"]
+            request_headers = self.request_headers()
+            request = self._create_prepared_request(
+                path=self.path(document_id=document_id),
+                headers=dict(request_headers, **self.authenticator.get_auth_header()),
+                params=self.request_params(),
+            )
+            response = self._send_request(request, {})
+            yield from self.parse_response(response)
+        else:
```

Comment:
why would we find ourselves in this situation? without any extra context I'm not sure that this isn't a silent failure

Reply:
Here we just decrypt and parse the report document. How it may be a silent failure?

Reply:
I meant the else statement -- the comment says "yield report if no document exists". If a document does not exist, then what is the report content? Isn't this a failure?

Reply:
@htrueman Because schemas impact the UX, I can't approve this PR until we resolve this question. Is the report payload the same schema as the report document? if not then we shouldnt output it because it will not be the schema that a user expects